src/cdk-experimental/selection/select-all.ts

```diff
@@ -0,0 +1,114 @@
+import {ControlValueAccessor, NG_VALUE_ACCESSOR} from '@angular/forms';
+import {BehaviorSubject, of as observableOf, Subject} from 'rxjs';
+import {switchMap, takeUntil} from 'rxjs/operators';
+
+import {CdkSelection} from './selection';
+
+/**
+ * Makes the element a select-all toggle.
+ *
+ * Must be used within a parent `CdkSelection` directive. It toggles the selection states
+ * of all the selection toggles connected with the `CdkSelection` directive.
+ * If the element implements `ControlValueAccessor`, e.g. `MatCheckbox`, the directive
+ * automatically connects it with the select-all state provided by the `CdkSelection` directive. If
+ * not, use `checked$` to get the checked state, `indeterminate$` to get the indeterminate state,
+ * and `toggle()` to change the selection state.
+ */
+@Directive({
+  selector: '[cdkSelectAll]',
+  exportAs: 'cdkSelectAll',
+})
+export class CdkSelectAll<T> implements OnDestroy, OnInit {
+  /**
+   * The checked state of the toggle.
+   * Resolves to `true` if all the values are selected, `false` if no value is selected.
+   */
+  readonly checked$ = new BehaviorSubject(false);
+
+  /**
+   * The indeterminate state of the toggle.
+   * Resolves to `true` if part (not all) of the values are selected, `false` if all values or no
+   * value at all are selected.
+   */
+  readonly indeterminate$ = new BehaviorSubject(false);
+
+  /**
+   * Toggles the select-all state.
+   * @param event The click event if the toggle is triggered by a (mouse or keyboard) click. If
+   *     using with a native <input type="checkbox">, the parameter is required for the
+   *     indeterminate state to work properly.
+   */
+  toggle(event?: MouseEvent) {
+    // This is needed when applying the directive on a native <input type="checkbox">
+    // checkbox. The default behavior needs to be prevented in order to support the indeterminate
+    // state. The timeout is also needed so the checkbox can show the latest state.
+    if (event) {
+      event.preventDefault();
+    }
+
+    setTimeout(() => {
+      this._selection.toggleSelectAll();
+    });
+  }
+
+  private readonly _destroyed$ = new Subject();
+
+  constructor(
+      @Optional() private readonly _selection: CdkSelection<T>,
+      @Optional() @Self() @Inject(NG_VALUE_ACCESSOR) private readonly _controlValueAccessor:
+          ControlValueAccessor[]) {}
+
+  ngOnInit() {
+    if (!this._selection) {
+      throw new Error('CdkSelectAll: missing CdkSelection in the parent');
+    }
+
+    if (!this._selection.cdkSelectionMultiple) {
+      throw new Error('CdkSelectAll: CdkSelection must have cdkSelectionMultiple set to true');
+    }
+
+    this._selection.cdkSelectionChange
+        .pipe(
+            switchMap(() => observableOf(this._selection.isAllSelected())),
+            takeUntil(this._destroyed$),
+            )
+        .subscribe((state) => {
+          this.checked$.next(state);
+        });
```

Comment:
Subscribing and then emitting is generally something you want to avoid. In this case, you can simply do
```ts
checked = this._selection.change.pipe(...);
```
I think you can avoid setting this up in `ngOnInit` too if the stream in `CdkSelection` is set up to support it

(here and elsewhere)